### PR TITLE
Testing other clouds in the master branch

### PR DESCRIPTION
**Pull request by danieljurek**
@weshaggard  -- Here's a proposal for the architecture using cloud configurations to multiply the matrix. 

There are two big problems I'd like to sort out: 

* Setting up environment-specific environment variables (for 2 definitions of "environment") 
* `ArmTemplateParameters` can't be splatted into the top level `@subscriptionConfiguration` so they'll have to go somewhere else

### CODE REVIEW

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:60]: This is an good starting point and will enable our scheduled runs but we might still need to figure out how we can allow people to trigger a test run for a particular cloud.

**Comment by weshaggard** [comment]
This definitely looks like a reasonable starting point.

File: eng/pipelines/templates/jobs/archetype-sdk-tests-jobs.yml
Comment (heaths) [inline eng/pipelines/templates/jobs/archetype-sdk-tests-jobs.yml:32]: Can we keep the names the same as they are used everywhere else, like azure-cli, e.g. `AzureCloud`? It'll be more user-friendly.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (heaths) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:45]: As above, I recommend `AzureCloud`.

**Review by heaths — APPROVED** [review]

**Review by heaths — APPROVED** [review]

File: eng/common/TestResources/New-TestResources.ps1
Comment (weshaggard) [inline eng/common/TestResources/New-TestResources.ps1:518]: FYI this touches eng/common so will need to be updated with tools process.

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:518]: I only proposed "AzureCloud" since that's how it's referred to in other places. If that's too risky given the downstream changes required, please disregard.

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:518]: I think it's just the parameter documentation that's out of date, not any functionality (actual values default to `AzureCloud`). Fix: https://github.com/Azure/azure-sdk-tools/pull/1127

**Comment by benbp** [comment]
Squashed/rebased this against master. @weshaggard I ran 6 different (pseudo-randomly selected) live test pipelines from various SDKs. There have been some timeout issues but so far they all look related to today's devops availability issues, and pass after retry. The failed tests currently linked are a result of me accidentally running the non-live test pipelines against this branch :/ I'm planning on a little more validation, but at this point I'm feeling decent about the safety of these changes as is.

**Comment by benbp** [comment]
Looks like the pipelines got unlinked.

https://dev.azure.com/azure-sdk/internal/_build/results?buildId=583059&view=results
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=583120&view=results
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=583121&view=results
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=583123&view=results
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=583125&view=results
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=583154&view=results

**Review by weshaggard — APPROVED** [review]
This change looks good from as starting point. As you mentioned yesterday we should do some follow-ups to try and start enabling some of these other clouds. Perhaps starting with the smoke-tests or some of the other live test pipelines that need to test in other clouds like Canary.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (weshaggard) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml:58]: We should decide if we want to include our other subscription configurations as part of these. Things like canary and preview.

Comment (benbp) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml:58]: From a pipeline authoring/simplicity perspective, I'm leaning towards representing the subscription configurations within their own cloud configs, as opposed to injecting them in via boolean parameters and if conditionals. This way as we add more configurations, it just becomes an entry to the cloud config array, meaning all combinations get captured explicitly by name.

Right now there are a few sdk configs consuming the SubscriptionConfiguration parameter directly, so it shouldn't be hard to move one way or the other as far as keeping sdk tests working.

**Comment by benbp** [comment]
> This change looks good from as starting point. As you mentioned yesterday we should do some follow-ups to try and start enabling some of these other clouds. Perhaps starting with the smoke-tests or some of the other live test pipelines that need to test in other clouds like Canary.

I'm starting to think getting sovereign cloud working is going to require sdk changes in a few places, so I agree with canary/preview/smoke tests as the starting point.

**Review by mitchdenny — COMMENTED** [review]
I'm happy with this approach for now. I think it is good to be able to trigger the pipelines via the UI and pick the clouds you want to run against. However before we merge this what consideration do we want to give to having the default be run all clouds? (guess I am echoing Wes' comment there).

Did you solve the concurrency issue too (i.e. limiting the number of concurrent agents consumed). Cosmos for example might cause a fairly large blow out in jobs.

**Comment by weshaggard** [comment]
Yes before we start enabling all the clouds in any larger set of pipelines we need to figure out what we can do about limiting the number of agents a new pipeline run picks up and uses at a time. From a quick look I didn't find any way to limit the number of parallel jobs for a pipeline similar to what you can do with a job matrix. That will definitely become a problem we have to solve. 

Another thing that I was also just thinking about with this approach is that it assumes all the platform configurations are suppose to run in each cloud (i.e. the cloud is the outer loop) so we might need to figure out some way to allow a platform to opt-out of  a particular cloud type as well. Similar to what we did with the "Record" condition.

**Comment by benbp** [comment]
Agree with both comments @mitchdenny @weshaggard. My plan was to:

1. First merge this PR with Daniel's changes while the sdk tests are green, so we can keep each update small/reviewable.
2. Start a new review with my changes to use the stage model (https://github.com/benbp/azure-sdk-for-net/tree/benbp/cloudconfig-template-test). This should make it a little easier to tune the platform and stage/cloud configurations independently, as well as be able to enable/disable clouds per run.
3. Start the process to enable new cloud/environment runs by default (canary, etc.)
4. Start to iterate on concurrency parameters, and on sparse matrix definitions (Wes' point about platform per cloud).

So ideally this will be a series of 4+ PRs that are easy to review and revert.

**Pull request merged**
